plutus-chain-index/app/Indexer.hs

```diff
@@ -0,0 +1,107 @@
+import Control.Monad (forever, when)
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.List (findIndex)
+import Data.Map (assocs)
+import Data.Maybe (fromJust, fromMaybe)
+import Data.Proxy (Proxy (Proxy))
+import Data.Text (pack)
+import Data.Text.Encoding (encodeUtf8)
+import Index.VSplit qualified as Ix
+import Ledger.TimeSlot (SlotConfig (..))
+import Plutus.ChainIndex.Tx (ChainIndexTx (..))
+import Plutus.Contract.CardanoAPI (fromCardanoTx)
+import Plutus.Script.Utils.V1.Scripts (Datum, DatumHash)
+
+-- We only care about the mainnet
+slotConfig :: SlotConfig
+slotConfig =
+  SlotConfig
+    { scSlotZeroTime = 1596059091000
+    , scSlotLength   = 1000
+    }
+
+networkId :: NetworkId
+networkId = Mainnet
+
+closeToGoguen :: ChainPoint
+closeToGoguen =
+  ChainPoint
+    (SlotNo 39795032)
+    (fromJust $ parseHash "3e6f6450f85962d651654ee66091980b2332166f5505fd10b97b0520c9efac90")
+
+parseHash :: String -> Maybe (Hash BlockHeader)
+parseHash hash =
+  deserialiseFromRawBytesHex (proxyToAsType Proxy) (encodeUtf8 $ pack hash)
+
+getDatums :: BlockInMode CardanoMode -> [(SlotNo, (DatumHash, Datum))]
+getDatums (BlockInMode (Block (BlockHeader slotNo _ _) txs) era) =
+  case era of
+    C.ByronEraInCardanoMode   -> concatMap (go era) txs
+    C.ShelleyEraInCardanoMode -> concatMap (go era) txs
+    C.AllegraEraInCardanoMode -> concatMap (go era) txs
+    C.MaryEraInCardanoMode    -> concatMap (go era) txs
+    C.AlonzoEraInCardanoMode  -> concatMap (go era) txs
+  where
+    go :: C.IsCardanoEra era
+       => C.EraInMode era C.CardanoMode
+       -> C.Tx era
+       -> [(SlotNo, (DatumHash, Datum))]
+    go era' tx =
+      let hashes = either (const []) (assocs . _citxData) $ fromCardanoTx era' tx
+      in  map (slotNo,) hashes
+
+processBlock :: IORef DatumIndex -> ChainSyncEvent -> IO ()
+processBlock ixref = \case
+  -- Not really supported
+  Resume point         -> putStrLn ("resume " <> show point) >> pure ()
+  RollForward blk@(BlockInMode (Block (BlockHeader slotNo _ blockNo@(BlockNo b)) _txs) _era) _tip -> do
+    when (b `rem` 1000 == 0) $
+      putStrLn $ show slotNo <> " / " <> show blockNo
+    ix  <- readIORef ixref
+    ix' <- Ix.insert (getDatums blk) ix
+    writeIORef ixref ix'
+  RollBackward point tip -> do
+    putStrLn ("rollback to " <> show tip)
+    rollbackToPoint point ixref
+
+rollbackToPoint
+  :: ChainPoint -> IORef DatumIndex -> IO ()
+rollbackToPoint point ixref = do
+  ix     <- readIORef ixref
+  events <- Ix.getEvents (ix ^. Ix.storage)
+  let ix' = fromMaybe ix $ rollbackOffset events ix
+  writeIORef ixref ix'
+  where
+    rollbackOffset :: [Ix.Event] -> DatumIndex -> Maybe DatumIndex
+    rollbackOffset events ix = do
+      slot   <- chainPointToSlotNo point
+      offset <- findIndex (any (\(s, _) -> s < slot)) events
+      Ix.rewind offset ix
+
+main :: IO ()
+main = do
+  tix <- Ix.open "datum.sqlite" 2160 >>= newIORef
+  _ <- runChainSync "/tmp/node.sock"
```

Comment:
We need to be able to pass a different location of the node socket.

Reply:
Do you still think we need to do this, if we just use the client until `plutus-streaming` gets merged?

Reply:
I am not expecting the user to having to change their node configuration to use this indexer. Yes, I don't think hardcoding the socket path is a good idea. (PS, I just merged #476)